### PR TITLE
RFC - PR - Request bound, early hinted assets

### PR DESCRIPTION
**Pull request by digitalsadhu**
~~I've spent some time considering how best to send assets, request bound, from a podlet to a layout. I wanted to make sure it was backwards compatible and didn't interfere with existing mechanisms as much as possible. I also wanted to see if early hinting could be used to get assets to the browser faster. With these things in mind, this PR sketches out how I think we could achieve these things from the podlet side. I've commented inline but as an overview, this is how a developer would code their podlet content hook:~~

Reworked based on feedback:

1. As Trygve mentions, Node v18.11.0 only for this feature.
2. The asset api surface remains the same and early hints are sent in the process method behind the scenes. Apart from point 1 above, those won't affect users.
3. Once we are happy with this, I will rebase the commit messages to reflect the changes

### CODE REVIEW

**Review by wkillerud — COMMENTED** [review]
Looks promising 👍 

I see the hints can include information like a [content security policy](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/103#early_hint_response_with_csp). Would it make sense to add a `csp` as a second optional argument?

I also wonder if we should name the method `res.hints` or `res.earlyHints` to be closer to the name of the response type, in case stuff gets added to it in the future.

**Comment by digitalsadhu** [comment]
> Looks promising 👍
> 
> I see the hints can include information like a [content security policy](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/103#early_hint_response_with_csp). Would it make sense to add a `csp` as a second optional argument?

I don't think it makes sense to send CSP information with 103 early hints. As I understand it, CSP is not enforced by the browser during the early hinting and if you had multiple podlets sending 103 early hints and this could result in multiple conflicting CSP definitions from different podlets. A single CSP definition should really come from the layout with the final payload.

> 
> I also wonder if we should name the method `res.hints` or `res.earlyHints` to be closer to the name of the response type, in case stuff gets added to it in the future.

I see your point for sure but as I see it, early hints is more of the underlying implementation and the feature we want to solve here is how to get assets from the podlet to the layout request bound. My naming with .assets might well be off though and I'm actually wondering if sticking to .js and .css might be best. Originally, I just assumed we could only send 1x a 103 early hint but you can actually send multiple so we could use .js and .css as separate methods.

**Comment by trygve-lie** [comment]
I don't think we need a separate method like this:

```js
    res.assets([
        { value: 'https://example.com/style.css' }
        { value: 'https://example.com/script.js' }
    ]);
```

How Podlets are today they have one set of asses which is set pr Podlet. In other words the assets is static during the life time of the server. I think we can just send off the Earl Hint with the assets in the `.process()` (the method called in `.middleware` in Express and in the Fastify plugin) as early as possible. Then there is no need for developers to call anything extra to send this off.

I think that in the [CSS](https://github.com/podium-lib/utils/blob/main/lib/asset-css.js) and [JS](https://github.com/podium-lib/utils/blob/main/lib/asset-js.js) asset classes we add a `.toHint()` method which returns a Early Hint Header representation of the Asset Object. Something like so:

```js
const css =  new PodiumAssetCss({ value: '/my-style.css' });
css.toHint();    // returns: '</my-style.css>; rel=preload; as=style'

const js =  new PodiumAssetJs({ value: '/my-script.js' });
js.toHint();    // returns: '<//my-script.js>; rel=preload; as=script'
```

Then, in the `.process()`:

```js
const earlyHintsLinks = [];

// Loop over css objects and build Early Hint links;
this.cssRoute.forEach((obj) => {
    earlyHintsLinks.push(obj.toHint());
});

// Loop over js objects and build Early Hint links;
this.jsRoute.forEach((obj) => {
    earlyHintsLinks.push(obj.toHint());
});

incoming.response.writeEarlyHints({
  'link': earlyHintsLinks,
});
```

**Comment by digitalsadhu** [comment]
> I don't think we need a separate method like this:
> 
> ```js
>     res.assets([
>         { value: 'https://example.com/style.css' }
>         { value: 'https://example.com/script.js' }
>     ]);
> ```
> 
> How Podlets are today they have one set of asses which is set pr Podlet. In other words the assets is static during the life time of the server. I think we can just send off the Earl Hint with the assets in the `.process()` (the method called in `.middleware` in Express and in the Fastify plugin) as early as possible. Then there is no need for developers to call anything extra to send this off.

Except that we need to support multi tenanting and this means that developers need to be able to set the assets being sent in the route hook... after the middleware has already run. 

> I think that in the [CSS](https://github.com/podium-lib/utils/blob/main/lib/asset-css.js) and [JS](https://github.com/podium-lib/utils/blob/main/lib/asset-js.js) asset classes we add a `.toHint()` method which returns a Early Hint Header representation of the Asset Object. Something like so:

Agree here for sure, I was thinking the same. 


> ```js
> const css =  new PodiumAssetCss({ value: '/my-style.css' });
> css.toHint();    // returns: '</my-style.css>; rel=preload; as=style'
> 
> const js =  new PodiumAssetJs({ value: '/my-script.js' });
> js.toHint();    // returns: '<//my-script.js>; rel=preload; as=script'
> ```
> 
> Then, in the `.process()`:
> 
> ```js
> const earlyHintsLinks = [];
> 
> // Loop over css objects and build Early Hint links;
> this.cssRoute.forEach((obj) => {
>     earlyHintsLinks.push(obj.toHint());
> });
> 
> // Loop over js objects and build Early Hint links;
> this.jsRoute.forEach((obj) => {
>     earlyHintsLinks.push(obj.toHint());
> });
> 
> incoming.response.writeEarlyHints({
>   'link': earlyHintsLinks,
> });
> ```

The challenge we have is that we are going to want the assets to be changed from request to request for multi tenanted apps. This means:

1. we need a way for the developer to reset the assets cache as part of or before calling .js and .css in a request hook.
2. we then need a way for the developer to send the assets.

One option would be to keep the api surface pretty similar and add a clearAssetCache method. .js and .css would still not have access to the res object so something would have to be added to send the hints.

Something like:

```js
app.get("/", (req, res) => {
  podlet.resetAssetCache();
  podlet.js(...);
  podlet.css(...);
  res.sendAssetEarlyHints();

  res.podiumSend(...)
});
```

or we could just add res.js and res.css instead that do early hinting as part of things:

```js
app.get("/", (req, res) => {
  res.js(...);
  res.css(...);

  res.podiumSend(...)
});
```

Thoughts?

**Comment by trygve-lie** [comment]
> Except that we need to support multi tenanting 

Right. Then I think this is about moving the process of setting asset in the podlet request bound and not so much about setting Early Hints. From the developers perspective I think its no need to know we do this with Early Hints. 

Then I think we can do something like so as you suggest:

```js
app.get("/", (req, res) => {
  res.js(...);
  res.css(...);

  res.podiumSend(...)
});
```

An then we do this inside `.podiumSend()`;

```js
const earlyHintsLinks = [];

// Loop over css objects and build Early Hint links;
incoming.css.forEach((obj) => {
    earlyHintsLinks.push(obj.toHint());
});

// Loop over js objects and build Early Hint links;
incoming.js.forEach((obj) => {
    earlyHintsLinks.push(obj.toHint());
});

incoming.response.writeEarlyHints({
  'link': earlyHintsLinks,
});
```

**Comment by digitalsadhu** [comment]
@trygve-lie and I took this discussion a bit offline so here's a summary of what we decided.

1. It makes sense to apply early hints as early as possible (in middleware) in order to avoid any slow work (that could also happen in middleware) and to avoid the developer needing to know anything about the early hint api etc.
2. We have a use case internally where apps are multi-tenanted and need different CDN URLs for different brands. To solve this problem, another Podium feature should implemented, context parser plugins in podlets. Once thats in place, a custom context parser for our podlets could be written to use brand to set cdn urls.

With that in mind, I will rework this PR to scope it to point 1. above.

File: lib/podlet.js [hunk 1]
```diff
@@ -863,11 +876,26 @@ export default class PodiumPodlet {
...
+        // @ts-expect-error
+        const js = incoming.js.map((asset) => asset.toHeader());
+
+        // @ts-expect-error
+        const css = incoming.css.map((asset) => asset.toHeader());
+
+        // Send early hints to layout client in the form of a 103 status code and a link header with js/css asset information.
+        // Always send this. If no assets present, send an empty string.
+        if (incoming.response.writeEarlyHints) {
+            incoming.response.writeEarlyHints({
```
Comment (wkillerud) [inline lib/podlet.js:894]: ❤️

**Review by wkillerud — APPROVED** [review]
I opened two PRs to sort out the JSDoc type stuff. This looks really good IMO 👍

**Comment by digitalsadhu** [comment]
Creating an early release by merging into next

**Pull request merged**